---
title: DeLorean analysis of McDavid et al. cell cycle data
author: John Reid
bibliography: DeLorean.bib
output:
  html_document:
    toc: true
    fig_caption: true
    # css: ../inst/Rmd/foghorn.css
---

<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{DeLorean analysis of McDavid et al. cell cycle data}
-->


```{r build, echo=FALSE, eval=FALSE}
library(devtools)
load_all('..')
library(rmarkdown)
render('McDavid-DeLorean.Rmd')

```


```{r config, echo=FALSE, message=FALSE}
library(knitr)
library(knitcitations)
library(rmarkdown)
#
# knitr options
#
opts_chunk$set(
    fig.path = 'figures/McDavid-',
    stop_on_error = TRUE,
    fig.width = 12.5,
    fig.height = 8)
#
# Citations
#
cleanbib()
cite_options(
    # hyperlink = 'to.doc',
    hyperlink = TRUE,
    # style = 'html',
    # citation_format = 'text',
    citation_format = "pandoc",
    cite.style = "numeric",
    check.entries = TRUE)
    # hyperlink = TRUE)
bib <- read.bibtex("DeLorean.bib")
if (file.exists("config.R")) {
    source("config.R")
}
#
# Widths for saving figures
#
text.width <- 4.7  # LaTeX width in inches
golden.ratio <- 1.618  # Pleasing ratio
fig.width <- text.width
fig.height <- text.width / golden.ratio

```

```{r init, echo=FALSE, message=FALSE, warning=FALSE}
# suppressMessages(loadfonts())
library(DeLorean)
#
# Stylesheet
#
options(markdown.HTML.stylesheet = system.file("inst/Rmd/foghorn.css",
                                               package="DeLorean"))
font.family <- "Verdana"
font.theme <- theme_update(text=element_text(family=font.family))
theme_set(font.theme)

```


`r citet(bib[["mcdavid_modeling_2014"]])` assayed actively cycling single
cells in order to examine the confounding effect of the cell cycle on
single cell expression analyses. They measured the expression of 333 genes
in 930 cells across three cell cycle phases and three cell lines.


# Data

McDavid et al.'s data is available in the `DeLorean` R package. Not all 333
genes are represented as several are rarely expressed.
```{r loadLib}
library(DeLorean)
data(McDavidDeLorean)

```

Shift positively expressed genes closer to unexpressed genes to improve fit.
```{r}
dl <- de.lorean(
    pmax(mcdavid.expr - .69, 0),
    mcdavid.gene.meta,
    mcdavid.cell.meta)

```

```{r child=report.file("data")}
```



# PCA analysis

```{r}
genes.high.rank <- (
    dl$gene.meta
    %>% filter(!is.na(cbRank), cbRank < 201)
    %>% arrange(cbRank))
levels(dl$cell.meta$cellline)
cells.PC3 <- mcdavid.cell.meta %>% filter("PC3" == cellline)
pca <- prcomp(t(mcdavid.expr[as.character(genes.high.rank$gene),
                             as.character(cells.PC3$cell)]),
              .scale=TRUE)
# Sqrt of eigenvalues
qplot(pca$sdev)
# Percentage of variance explained
qplot(pca$sdev**2 / sum(pca$sdev**2) * 100)
# Get the PC scores
pc.scores <- as.data.frame(pca$x)
pc.scores$cell <- factor(rownames(pca$x), levels=levels(dl$cell.meta$cell))
pc.scores <- pc.scores %>% left_join(dl$cell.meta)
# Plot PC1 against PC2
ggplot(pc.scores, aes(x=PC1, y=PC2, color=capture)) + geom_point()
# Plot PC1 against PC3
ggplot(pc.scores, aes(x=PC1, y=PC3, color=capture)) + geom_point()

```


# Choose genes and cells

Choose a few cells from the PC3 cell line.
```{r filterCells}
dl <- filter.cells(dl, cells=cells.PC3$cell)
set.seed(getOption("McDavid.seed", 1))
max.cells <- min(getOption("McDavid.max.cells", 17))
dl <- filter.cells(dl, number=max.cells)

```


Choose genes with low $p$-values from the McDavid et al.
differential expression test.
```{r filterGenes}
max.genes <- min(getOption("McDavid.max.genes", 21))
genes.for.stan <- (
    dl$gene.meta
    %>% filter(pvalue > 10)
    %>% sample_n(max.genes))
dl <- filter.genes(dl, genes=genes.for.stan$gene)

```


# Estimate hyperparameters

Examine data for empirical Bayes estimation of hyperparameters.
```{r empiricalBayes}
dl <- adjust.by.cell.sizes(dl)
ggplot(dl$cell.sizes, aes(x=capture, y=size)) + geom_boxplot()
dl <- estimate.hyper(
    dl,
    sigma.tau=.5,
    length.scale=5,
    delta=.5,
    model.name=getOption("McDavid.model", 'simple-model'))

```

```{r child=report.file("hyper-parameters")}
```


Save expression data and meta data.
```{r}
saveRDS(list(expr = dl$expr, cell.meta = dl$cell.map, gene.meta=dl$gene.map),
        file='McDavid-input.rds')

```


# Compile and fit model

Define and compile the model, find the best initialisation, and fit the model.
```{r model}
dl <- format.for.stan(dl, period=3, hold.out=3)
dl <- compile.model(dl)
dl <- find.best.tau(dl)
# dl <- find.smooth.tau(dl)
system.time(dl <- fit.model(dl))

```


# Examine convergence.

```{r examConv}
dl <- examine.convergence(dl)

```

```{r child=report.file("convergence")}
```



# Analyse posterior

Examine posterior.
```{r posterior}
dl <- process.posterior(dl)
# dl <- optimise.best.sample(dl)
dl <- analyse.noise.levels(dl)

```
```{r child=report.file("posterior")}
```


# Profiles

Calculate expression profiles.
```{r makePredictions}
dl <- make.predictions(dl)
```
```{r child=report.file("profiles")}
```


# Examine peak times

```{r peakTimeFns}
# A function to make times periodic
periodise <- function(tau, period=3) {
    wave = floor(tau/period)
    tau - wave * period
}
# tau.test <- seq(-.5, 3.5, by=.5)
# tau.test
# periodise(tau.test)
# Set up some functions to map between pseudotimes and peak times
#
# cbPeaktimes of 0 (equiv. 100) correspond to the transition
# between M and G1.
# (http://cyclebase2.jensenlab.org/methods.jsp#peaktime-uncertainty)
# Hence cbPeaktime = 0 corresponds to a pseudotime of .5
# A function to map from pseudotimes (tau) to cbPeaktimes
tau.to.cbtime <- function(tau) 100*periodise((tau - .5)/3, 1)
# and vice versa
cbtime.to.tau <- function(cbtime) periodise(cbtime/100*3+.5, 3)
# Plot cbPeaktime against best sample max predicted mean time
# print(plot.profiles(dl, genes="CDCA8"))
#
# Distance between two points on interval [0, 100] where 0 is connected to 100
peak.distance <- function(peak.1, peak.2) {
    stopifnot(all(peak.1 >= 0.))
    stopifnot(all(peak.1 <= 100.))
    stopifnot(all(peak.2 >= 0.))
    stopifnot(all(peak.2 <= 100.))
    dist <- abs(peak.1 - peak.2)
    ifelse(dist > 50, 100 - dist, dist)
}
# Test peak distances
# peak.distance(c(.1, .2, .1, .9, .5, .5), c(.2, .1, .9, .1, .5, .4))
#
# RMSE functions
#
# Calculate the root mean square of d
calc.rms <- function(d) sqrt(sum(d**2))/length(d)
# Calculate the root mean square error between xhat and x.
calc.rmse <- function(xhat, x) calc.rms(peak.distance(xhat, x))

```


Calculate distance between peak time and capture time.
Capture times are as defined in
[CycleBase](http://www.cyclebase.org/About "CycleBase").
```{r peakTimeDistances}
peaks <- (
    melt.expr(dl)
    %>% left_join(dl$cell.map %>% select(cell, capture, obstime))
    %>% group_by(gene)
    %>% summarise(peak.idx=which.max(x),  # Find index of maximal expression
                  cell=cell[peak.idx],
                  capture=capture[peak.idx],
                  obstime=obstime[peak.idx])
    %>% left_join(dl$gene.map %>% select(gene, cbPeaktime))
    %>% filter(! is.na(cbPeaktime))
    %>% mutate(capture.dist=peak.distance(cbPeaktime,
                                          tau.to.cbtime(obstime))))
peaks
# Check peak times are correct with random gene.
gene.test <- sample(rownames(dl$expr), 1)
gene.test
gene.map.test <- filter(dl$gene.map, gene.test == gene)
expr.test <- dl$expr[gene.test,]
expr.test
which.max(expr.test)
cell.max <- names(which.max(expr.test))
cell.max
cell.map.test <- filter(dl$cell.map, cell.max == cell)
cell.map.test
filter(peaks, gene.test == gene) %>% select(cbPeaktime,
                                            cell, capture, capture.dist)
stopifnot(cell.max == filter(peaks, gene.test == gene)$cell)
# Calculate RMSE for naive method
capture.RMSE <- calc.rms(peaks$capture.dist)
capture.RMSE

```


```{r predictedPeaks}
dl <- within(dl, {
    # Find the peaks in the predicted profiles
    peak.vs.max <- (
        predictions
        %>% group_by(iter, g)
        %>% summarise(predicted.peak=tau.to.cbtime(tau[which.max(predictedmean)]))
        %>% left_join(gene.map)
        %>% filter(! is.na(cbPeaktime)))
    # Calculate the distance between the peaks and the CycleBase peaktimes
    peak.max.dist <- (
        peak.vs.max
        %>% group_by(iter)
        %>% summarise(rmse=calc.rmse(cbPeaktime/100,
                                     predicted.peak)))
})
best.peak.dist <- with(dl, filter(peak.max.dist, best.sample == iter)$rmse)
sample_n(dl$peak.max.dist, 15)
# Plot distribution
gp.RMSE <- (
    ggplot(dl$peak.max.dist, aes(x=rmse))
    + geom_density()
    + geom_rug()
    + geom_vline(x=capture.RMSE, linetype='dashed', color='red')
    + geom_vline(x=best.peak.dist, linetype='dashed', color='blue')
    + scale_x_continuous(name="RMSE"))
ggsave('McDavid-rmse.pdf', plot=gp.RMSE,
       width=text.width, height=text.width)
ggsave('McDavid-rmse.png', plot=gp.RMSE,
       width=fig.width*1.75, height=fig.height*1.5,
    units="in", dpi=200)
print(gp.RMSE)
# Mean of RMSE
posterior.RMSE.mean <- mean(dl$peak.max.dist$rmse)
posterior.RMSE.mean
gp.peak.vs.max <- (
    ggplot(dl$peak.vs.max %>% filter(dl$best.sample == iter),
           aes(x=cbPeaktime, y=predicted.peak, label=gene))
    + geom_point(alpha=.7, size=2)
    + scale_x_continuous(name="CycleBase peak time")
    + scale_y_continuous(name="estimated peak time")
)
print(gp.peak.vs.max)
ggsave('McDavid-peak-vs-max.pdf', plot=gp.peak.vs.max,
       width=text.width, height=text.width)
ggsave('McDavid-peak-vs-max.png', plot=gp.peak.vs.max,
       width=fig.width*1.75, height=fig.height*1.5,
       units="in", dpi=200)

```

```{r testPeaks}
# Test for normality
with(dl$peak.max.dist, shapiro.test(rmse))
gp.qq <- (
    ggplot(dl$peak.max.dist, aes(sample=rmse))
    + stat_qq())
print(gp.qq)
ggsave('McDavid-rmse-qq.pdf', plot=gp.qq,
       width=text.width, height=text.width)
# Assuming a normal distribution for the posterior RMSE, what is
# the likelihood of observing at least as extreme as the
# peak capture RMSE.
with(dl$peak.max.dist,
     1-pnorm(capture.RMSE, mean=mean(rmse), sd=sd(rmse), lower.tail=F))
gp.capture.peaks <- (
    ggplot(peaks, aes(x=capture, y=cbPeaktime))
    + geom_boxplot()
    + scale_y_continuous(name="CycleBase peak time")
    + scale_x_discrete(name="cell cycle phase")
    + coord_flip())
print(gp.capture.peaks)
ggsave('McDavid-capture-peaks.pdf', plot=gp.capture.peaks,
       width=text.width, height=text.width)
ggsave('McDavid-capture-peaks.png', plot=gp.capture.peaks,
       width=fig.width*3.5, height=fig.height*3,
       units="in", dpi=100)

```


Plot the profiles adjusted for the cell size.
```{r}
genes.for.profiles <- with(dl,
    samples.l$psi
    %>% filter(best.sample == iter)
    %>% left_join(samples.l$omega)
    %>% arrange(-psi/omega)
    %>% head(12)
    %>% left_join(gene.map)
)
add.peak.times <- function(dl, gp) {
    plot.peaktimes <- (
        dl$gene.meta
        %>% filter(gene %in% gp$data[['gene']], ! is.na(cbPeaktime))
        %>% mutate(peak.tau=cbtime.to.tau(cbPeaktime))
    )
    (
        gp
        + geom_vline(data=plot.peaktimes,
                     aes(xintercept=peak.tau),
                     linetype='dashed', alpha=.7)
    )
}
gp.profiles <- (
    plot(dl, type="profiles", genes=genes.for.profiles$gene)
    + scale_x_continuous(name='Pseudotime', limits=c(0, 3))
)
gp.profiles <- add.peak.times(dl, gp.profiles)
print(gp.profiles)
ggsave('McDavid-profiles.png', plot=gp.profiles,
    width=fig.width*3.5, height=fig.height*3,
    units="in", dpi=100)
ggsave('McDavid-profiles.pdf', plot=gp.profiles,
       width=2*text.width, height=2*text.width/golden.ratio)

```


```{r echo=FALSE}
# Save DeLorean object without fit component
saveRDS({dl2 <- dl; dl2$fit <- NULL; dl2}, "McDavid.rds")
```


```{r date}
date()
```
R version and packages used:
```{r Rversion}
sessionInfo()
```
