```{r hyperAttach, include=FALSE}
# Attach the DeLorean data frame to access members
attach(dl)
```

## Estimation of cell size factors: $S_c$

The cell size factors represent the relative amount of biological material in
each cell. They are used to account for effects such as sequencing depth or
lysis efficiency.
First we calculate the mean expression for each gene, that is the average of
all its expression values.
$$
\mu_g = \textrm{Mean}_c (x_{g,c})
$$
which has this density over genes:
```{r geneMean}
(ggplot(gene.expr, aes(x=x.mean)) + geom_density() + geom_rug())

```

Now we estimate the cell size factors. We use the median for robustness and
examine every gene's relative expression levels in each cell
$$
\hat{S}_c = \text{Median}_g (x_{g,c} - \mu_g)
$$
We estimate the hyperparameters for the cell size factors by fitting a normal
distribution. The estimated cell size factors have this density
(we show the empirical Bayes prior in dotted blue):
```{r cellSizeFactors}
(ggplot(cell.expr, aes(x=S.hat)) + geom_density() + geom_rug()
    + stat_function(fun=function(x) dnorm(x,
                                          mean=hyper$mu_S,
                                          sd=hyper$sigma_S),
                    colour="blue", alpha=.7, linetype="dashed")
)

```



### Estimation of expression variation: $\phi_g$, $\psi_g$, $\omega_g$

Given the estimated cell size factors, we can adjust our gene
measurements, accounting for the different cell sizes
$$
\hat{x}_{g,c} = x_{g,c} - \hat{S}_c
$$

We expect this should reduce the variation in the expression levels on average.
We check this by plotting the standard deviation of the adjusted versus
the unadjusted.
```{r adjVsUnadj}
(ggplot(gene.expr,
        aes(x=x.sd, y=x.hat.sd))
    + geom_point()
    + geom_abline(intercept=0, slope=1, alpha=.4)
)

```


From these adjusted expression measurements, we can reestimate the genes' means
$$
\hat{\mu}_g = \textrm{Mean}_c (\hat{x}_{g,c})
$$
which have this density (again we show the hyperparameters as a
dotted blue line):
```{r geneAdjMean}
(ggplot(gene.expr, aes(x=phi.hat)) + geom_density() + geom_rug()
    + stat_function(fun=function(x) dnorm(x,
                                          mean=hyper$mu_phi,
                                          sd=hyper$sigma_phi),
                    colour="blue", alpha=.7, linetype="dashed")
)

```

Having estimated the gene expression means, we can decompose the observed
variance.  We examine the variance both between and within time points. We
expect some of the variance within a time point to be associated with noise in
the temporal dimension, that is estimating the pseudotime correctly should
reduce the within time variance. The estimation procedure has a parameter
$\delta$ (which in this example was `r opts$delta`) that allows the user to
determine how much of the within time variance they wish to ascribe to
between time variance.

Let
$$
k_c \in \{\kappa_1, \dots, \kappa_T\}
$$
be the time point at which cell $c$ was captured. We partition the cells
by their captured time points indexed by $1 \le t \le T$:
$$
\mathcal{K}_t = \{c: k_c = \kappa_t\}
$$
We group the expression measurements by gene and observed time to
calculate means and variances:
$$
\begin{align}
\mu_{g,t} &= \text{Mean}(\{\hat{x}_{g,c}:z_{g,c} = 1, c \in \mathcal{K}_t\}) \\\\
\mathbb{V}_{g,t} &= \text{Var}(\{\hat{x}_{g,c}:z_{g,c} = 1, c \in \mathcal{K}_t\})
\end{align}
$$
The variances have this distribution:
```{r geneWithinVar}
(ggplot(gene.time.expr, aes(x=x.var)) + geom_density() + geom_rug())

```

Mean expression by time:
```{r geneTimeMean}
(ggplot(gene.time.expr,
        aes(x=x.mean, y=sqrt(x.var), color=capture))
    + geom_point(alpha=.4)
)

```


We estimate the gene variation between time as the variation of the within
time mean:
$$
\bar{\psi}_g = \text{Var}_t(\mu_{g,t})
$$
and the gene variation within time as the average of the variation
within the times:
$$
\bar{\omega}_g = \text{Mean}_t(\mathbb{V}_{g,t})
$$

Gene variation between and within time:
```{r geneVarBetweenWithin}
(ggplot(gene.var,
        aes(x=omega.bar,
            y=psi.bar))
    + geom_point(alpha=.5)
    + geom_rug(alpha=.05)
)

```

We expect some of the variance within a time point to be associated with noise
in the temporal dimension so we adjust the estimated between and within
time variances using a scaling factor $\delta$:
$$
\begin{align}
\hat{\psi}_g &= \bar{\psi}_g + \delta \bar{\omega}_g \\\\
\hat{\omega}_g &= \bar{\omega}_g - \delta \bar{\omega}_g
\end{align}
$$
giving this density for the between time variation:
```{r geneVarBetweenAdj}
(ggplot(gene.var, aes(x=psi.hat)) + geom_density() + geom_rug()
    + stat_function(fun=function(x) dlnorm(x,
                                           mean=hyper$mu_psi,
                                           sd=hyper$sigma_psi),
                    colour="blue", alpha=.7, linetype="dashed")
)
```

and this density for the within time variation:
```{r geneVarWithinAdj}
(ggplot(gene.var, aes(x=omega.hat)) + geom_density() + geom_rug()
    + stat_function(fun=function(x) dlnorm(x,
                                           mean=hyper$mu_omega,
                                           sd=hyper$sigma_omega),
                    colour="blue", alpha=.7, linetype="dashed")
)

```

```{r hyperDetach, include=FALSE}
# Detach the previously attached DeLorean data frame
detach(dl)
```
